creduce/utils/testing.py

```diff
@@ -0,0 +1,847 @@
+
+    def kill(self, environments):
+        for test_env in environments:
+            #logging.debug("Kill {}".format(test_env.process_pid))
+
+            if not test_env.has_result() and not self.no_kill:
+                if sys.platform == "win32":
+                    self._kill_win32(test_env.process_pid)
+                else:
+                    self._kill_posix(test_env.process_pid)
+
+            test_env.wait_for_result()
+
+    @classmethod
+    def killpg(cls, pid):
+        if sys.platform == "win32":
+            cls._kill_win32(pid)
+        else:
+            cls._kill_posix(pid)
+
+
+class GeneralTestRunner(AbstractTestRunner):
+    def __init__(self, test_script, timeout, save_temps, no_kill):
+        super().__init__(test_script, timeout, save_temps, no_kill)
+
+        self.test_script = os.path.abspath(test_script)
+
+    @classmethod
+    def is_valid_test(cls, test_script):
+        for mode in {os.F_OK, os.X_OK}:
+            if not os.access(test_script, mode):
+                return False
+
+        return True
+
+    def create_environment(self):
+        return GeneralTestEnvironment(self.test_script, self.timeout, self.save_temps)
+
+class PythonTestRunner(AbstractTestRunner):
+    def __init__(self, test_module, timeout, save_temps, no_kill):
+        super().__init__(test_module, timeout, save_temps, no_kill)
+
+        if os.path.isfile(test_module):
+            (module_name, _) = os.path.splitext(os.path.basename(test_module))
+            self.module_spec = importlib.util.spec_from_file_location(module_name, os.path.abspath(test_module))
+        else:
+            self.module_spec = importlib.util.find_spec(test_module)
+
+    @classmethod
+    def is_valid_test(cls, test_script):
+        try:
+            if os.path.isfile(test_script):
+                (module_name, _) = os.path.splitext(os.path.basename(test_script))
+                module_spec = importlib.util.spec_from_file_location(module_name, test_script)
+            else:
+                module_spec = importlib.util.find_spec(test_script)
+
+            if module_spec is None:
+                return False
+
+            module = importlib.util.module_from_spec(module_spec)
+            module_spec.loader.exec_module(module)
+        except FileNotFoundError:
+            return False
+
+        return (getattr(module, "run", None) is not None)
+
+    def create_environment(self):
+        return PythonTestEnvironment(self.module_spec, self.timeout, self.save_temps)
+
+class AbstractTestManager:
+    #TODO: How can we get this information?
+    PACKAGE = "TODO"
+    COMMIT = "TODO"
+    GIVEUP_CONSTANT = 50000
+    MAX_CRASH_DIRS = 10
+    MAX_EXTRA_DIRS = 25000
+    GIVEUP_CONSTANT = 50000
+
+    def __init__(self, test_runner, pass_statistic, test_cases, parallel_tests, no_cache, silent_pass_bug, die_on_pass_bug, print_diff, max_improvement, no_give_up, also_interesting):
```

Comment:
In line 139 in `creduce.py` (the top-level script) `args.n` is passed as argument to the constructor of the test manager.